examples/command.rs

```diff
@@ -0,0 +1,308 @@
+#![allow(clippy::type_complexity)]
+
+use valence::prelude::*;
+use valence::protocol::packets::play::command_tree_s2c::Parser;
+use valence_command::arg_parser::CommandArgParseError;
+use valence_command::command_graph::CommandGraphBuilder;
+use valence_command::command_scopes::CommandScopes;
+use valence_command::handler::{CommandExecutionEvent, CommandHandler};
+use valence_command::{arg_parser, Command, CommandArgSet, CommandScopeRegistry};
+
+const SPAWN_Y: i32 = 64;
+
+pub enum TeleportResult {
+    ExecutorToLocation(arg_parser::Vec3),
+    ExecutorToTarget(String),
+    TargetToTarget((String, String)),
+    TargetToLocation((String, arg_parser::Vec3)),
+}
+
+#[derive(Resource, Clone)]
+struct TeleportCommand;
```

Comment:
I don't think this struct needs to exist. Unless I'm missing something, all the methods get run once, so it seems that it would be better to make a function like

```rust
impl App {
    pub fn add_command<T>(&mut self, name: impl Into<String>, build: impl FnOnce(&mut CommandGraphBuilder<T>)) -> &mut Self;
}
```

Or to make this friendly for the derive macro, rename the above to `add_command_raw` and add

```rust
fn add_command<C: Command>(&mut self) -> &mut Self {
    self.add_command_raw(C::name(), C::assemble_graph)
}
```

`Command` would be implemented on `TeleportResult` in this example.

Reply:
I'd like the command resource to be capable of storing data for various reasons. I also want to be able to reassemble the graph at runtime by modifying the command resource. Remember I'm not done yet so I have not started to work on the final implementation.

Reply:
Right, I'm just trying the avoid the situation mentioned on Discord where the derive macro has to generate a secondary type that the user needs to know about.